tokio/src/blocking/mod.rs

```diff
@@ -0,0 +1,65 @@
+//! Perform blocking operations from an asynchronous context.
+
+mod pool;
+pub(crate) use self::pool::{BlockingPool, Spawner};
+
+mod schedule;
+mod task;
+
+use crate::task::JoinHandle;
+
+/// Run the provided blocking function without blocking the executor.
+///
+/// In general, issuing a blocking call or performing a lot of compute in a
+/// future without yielding is not okay, as it may prevent the executor from
+/// driving other futures forward.  If you run a closure through this method,
+/// the current executor thread will relegate all its executor duties to another
+/// (possibly new) thread, and only then poll the task. Note that this requires
+/// additional synchronization.
+///
+/// # Examples
+///
+/// ```
+/// # async fn docs() {
+/// tokio::blocking::in_place(move || {
+///     // do some compute-heavy work or call synchronous code
+/// });
+/// # }
+/// ```
+#[cfg(feature = "rt-full")]
+pub fn in_place<F, R>(f: F) -> R
+where
+    F: FnOnce() -> R,
+{
+    use crate::runtime::{enter, thread_pool};
+
+    enter::exit(|| thread_pool::block_in_place(f))
+}
+
+/// Run the provided closure on a thread where blocking is acceptable.
+///
+/// In general, issuing a blocking call or performing a lot of compute in a future without
+/// yielding is not okay, as it may prevent the executor from driving other futures forward.
+/// A closure that is run through this method will instead be run on a dedicated thread pool for
+/// such blocking tasks without holding up the main futures executor.
+///
+/// # Examples
+///
+/// ```
+/// # async fn docs() -> Result<(), Box<dyn std::error::Error>>{
+/// let res = tokio::blocking::spawn_blocking(move || {
+///     // do some compute-heavy work or call synchronous code
+///     "done computing"
+/// }).await?;
+///
+/// assert_eq!(res, "done computing");
+/// # Ok(())
+/// # }
+/// ```
+pub fn spawn_blocking<F, R>(f: F) -> JoinHandle<R>
```

Comment:
I'm on the fence about this but WDYT about just `blocking::spawn`? My expectation is that people will probably use the fns in this module as
```rust 
tokio::blocking::FN_NAME(...);
```
or 
```rust
use tokio::blocking; 

blocking::FN_NAME(...);
```
rather than
```rust
use tokio::blocking::FN_NAME;

FN_NAME(...);
```
so it seems a little weird to me to call `blocking::spawn_blocking`, esp. since the `in_place` fn does not have the word "blocking" in its name.

FWIW, with the local task stuff in #1733, I was going to call the spawn fn `local::spawn` when we decided to not re-export it from the root (although if it moves to `task`, which I expect it will after this PR merges, it will probably become `task::spawn_local`...

naming things is hard!

Reply:
I expect that we will move this to `task::spawn_blocking` and have all the misc spawning tasks in one location. Same w/ `in_place`.

Reply:
@carllerche okay, why not do that in this PR, then?

Reply:
it'll take more significant reorg of the current task module... and isn't critical for this PR which is already death.